### DISCUSSION TITLE
Singleton RLiveObjects

### DISCUSSION THREAD

**Discussion by yoursalvation** [redisson/redisson#6097] [Ideas]
Hey,
I would really like to see RLiveObject Singletons. In my mind they would act like similiarly to the RRemoteService, allowing to be registered behind a Class Type, whilst being stored and proxied by redisson instead of the local JVM. These would be quite useful, basically just being a RemoteService, but stored and synchronized by redisson.

This is useful in cases where a RemoteService is needed, but centralised in redis. I know that RBuckets exist, altough I'm not a big fan of them, as they aren't atomic / thread safe by nature like RLiveObjects.

Let me know what you guys think and thanks for reading.

**Comment by mrniko** [comment]
> I know that RBuckets exist, altough I'm not a big fan of them, as they aren't atomic / thread safe by nature like RLiveObjects.

What do you mean? RBuckets is also thread-safe object.

**Reply by yoursalvation** [reply]
Hmm, I might have used them wrong then. I tried before writing this and set up 2 Threads which got the RBucket, retrieved the object, incremented a number and set the RBucket to the new value.

Can you show me how they are supposed to be used?

**Reply by mrniko** [reply]
what was the issue? can you share the code?

**Reply by yoursalvation** [reply]
```
public class Main {

    private static RedissonClient redis;

    public static void main(String[] args) {
        redis = Redisson.create();

        RBucket<MyObject> bucket = redis.getConnection().getBucket("myBucket");
        bucket.set(new MyObject());


        new Thread() {
            @Override
            public void run() {
                for (int i = 0; i < 10000; i++) {
                    MyObject myObject = bucket.get();
                    myObject.incrementNumber();
                    bucket.set(myObject);
                }
                System.out.println("finished");
            }
        }.start();
        new Thread() {
            @Override
            public void run() {
                for (int i = 0; i < 10000; i++) {
                    MyObject myObject = bucket.get();
                    myObject.incrementNumber();
                    bucket.set(myObject);
                }
                System.out.println("finished");
            }
        }.start();

        try {
            Thread.sleep(10000);
        } catch (InterruptedException e) {

        }
        System.out.println(bucket.get().getNumber());

        //prints finished
        //prints finished
        //prints 10164 (should be 20000)

    }
}
```

```
public class MyObject {
    private int number;

    public MyObject() {
        number = 0;
    }

    public void incrementNumber() {
        this.number += 1;
    }

    public int getNumber() {
        return number;
    }
}
```

**Reply by mrniko** [reply]
How you did that code thread-safe with RLiveObject?

**Reply by yoursalvation** [reply]
I haven't actually tried with RLiveObjects (I was sure it would work, because the documentation says everything gets replaced with redis operations and redis is thread-safe) , but after trying it seems like it also return something around 10000 instead of the expected 20000.

**Reply by yoursalvation** [reply]
Are the fields in RLiveObjects not replaced by their RAtomic counter-part?

**Reply by mrniko** [reply]
No. Long type remains Long

**Reply by yoursalvation** [reply]
Okay, so how can I use the Bucket in an atomic fashion + is there a way for me to have RAtomic fields in RLiveObjects?

**Reply by yoursalvation** [reply]
Anyways, so what do you think of RLiveObject Singletons?

**Reply by mrniko** [reply]
>  is there a way for me to have RAtomic fields in RLiveObjects?

I'm afraid no.

> so what do you think of RLiveObject Singletons?

You can use it as singleton, since fields are backed by Redisson thread-safe objects.
